stateresolutionv2.go

```diff
@@ -0,0 +1,565 @@
+			break
+		}
+		// Membership events are only control events if the "membership" key in the
+		// content is "leave" or "ban" so we need to extract the content.
+		var content map[string]interface{}
+		if err := json.Unmarshal(e.Content(), &content); err != nil {
+			break
+		}
+		// If the "membership" key is set and is set to either "leave" or "ban" then
+		// the event is a control event.
+		if m, ok := content["membership"]; ok && (m == "leave" || m == "ban") {
+			return true
+		}
+	default:
+	}
+	// If we have reached this point then we have failed all checks and we don't
+	// count the event as a control event.
+	return false
+}
+
+// createPowerLevelMainline generates the mainline of power level events,
+// starting at the currently resolved power level event from the topological
+// ordering and working our way back to the room creation. Note that we populate
+// the result here in reverse, so that the room creation is at the beginning of
+// the list, rather than the end.
+func (r *stateResolverV2) createPowerLevelMainline() []Event {
+	var mainline []Event
+
+	// Define our iterator function.
+	var iter func(event Event)
+	iter = func(event Event) {
+		// Append this event to the beginning of the mainline.
+		mainline = append([]Event{event}, mainline...)
+		// Work through all of the auth event IDs that this event refers to.
+		for _, authEventID := range event.AuthEventIDs() {
+			// Check that we actually have the auth event in our map - we need this so
+			// that we can look up the event type.
+			if authEvent, ok := r.authEventMap[authEventID]; ok {
+				// Is the event a power event?
+				if authEvent.Type() == MRoomPowerLevels {
+					// We found a power level event in the event's auth events - start
+					// the iterator from this new event.
+					iter(authEvent)
+				}
+			}
+		}
+	}
+
+	// Begin the sequence from the currently resolved power level event from the
+	// topological ordering.
+	if r.resolvedPowerLevels != nil {
+		iter(*r.resolvedPowerLevels)
+	}
+
+	return mainline
+}
+
+// getFirstPowerLevelMainlineEvent iteratively steps through the auth events of
+// the given event until it finds an event that exists in the mainline. Note
+// that for this function to work, you must have first called
+// createPowerLevelMainline. This function returns three things: the event that
+// was found in the mainline, the position in the mainline of the found event
+// and the number of steps it took to reach the mainline.
+func (r *stateResolverV2) getFirstPowerLevelMainlineEvent(event Event) (
+	mainlineEvent Event, mainlinePosition int, steps int,
+) {
+	// Define a function that the iterator can use to determine whether the event
+	// is in the mainline set or not.
+	isInMainline := func(searchEvent Event) (bool, int) {
+		// First of all check the cache. If we've processed this event already then
+		// we should already know the mainline position.
+		if pos, ok := r.powerLevelMainlinePos[searchEvent.EventID()]; ok {
```

Comment:
This isn't quite what I meant with https://github.com/matrix-org/gomatrixserverlib/pull/153/files#r386698280 .  If just after https://github.com/matrix-org/gomatrixserverlib/pull/153/files#diff-23c7386d47026d5c47bccf815b1e6373R121 you do something like:

``` go
for pos, event := range r.powerLevelMainline {
  r.powerLevelMainlinePos[event] := pos
}
```

(sorry if my syntax is wrong) then the `isInMainline` function here just needs to be

``` go
if pos, ok := r.powerLevelMainlinePos[searchEvent.EventID()]; ok {
  return true, pos
} else {
  return false, 0
}
```

then you only need to loop through `r.powerLevelMainline`.  Whereas currently, you will loop through `r.powerLevelMainline` once every time you encounter an event that isn't in the mainline.

Reply:
Right, yes I see what you mean. I've done that now.